### PR TITLE
Refactor fallback' read header

### PR DESCRIPTION
**Pull request by GuyTuval**
I refactored the code in the `_read_header()` method of the `fallback.py` file.
I removed duplicated code and (hopefully) made the function more readable and maintainable.

### CODE REVIEW

**Comment by GuyTuval** [comment]
Hey @methane!
I saw that you are one of the contributors to this repo.
I wanted to ask if you would mind review this PR? 😄

**Comment by methane** [comment]
Would you try benchmark/benchmark.py ?

File: msgpack/fallback.py [hunk 1]
```diff
@@ -149,6 +149,45 @@ def _unpack_from(f, b, o=0):
     _unpack_from = struct.unpack_from
 
 
+class MessagePack:
```
Comment (jfolz) [inline msgpack/fallback.py:152]: Consider using a `namedtuple` instead.

**Comment by GuyTuval** [comment]
> 
> 
> Would you try benchmark/benchmark.py ?
> Run the script with master branch and your PR branch.
I ran the tests on my system.
The results with master branch:
```
packing integers (ext)           0.340488, 0.340425, 0.343065, 0.348732
packing integers (fallback)      4.452573, 4.616886, 4.559352, 4.515878
unpacking integers (ext)         0.066253, 0.066310, 0.068629, 0.069208
unpacking integers (fallback)    8.194786, 8.229670, 8.296369, 8.541644
packing bytes (ext)              0.034188, 0.034607, 0.034281, 0.034365
packing bytes (fallback)         0.737407, 0.747305, 0.775749, 0.839676
unpacking bytes (ext)            0.024286, 0.024064, 0.023509, 0.023577
unpacking bytes (fallback)       2.422867, 2.279277, 2.227503, 2.189443
packing lists (ext)              0.738786, 0.733464, 0.731821, 0.717495
packing lists (fallback)       10.754110, 10.833690, 11.016210, 10.988131
unpacking lists (ext)            0.287086, 0.288698, 0.291442, 0.286039
unpacking lists (fallback)     12.463816, 12.133081, 11.914829, 11.711606
packing dicts (ext)              0.438501, 0.456876, 0.459528, 0.449412
packing dicts (fallback)       12.814651, 12.784222, 12.757698, 12.723056
unpacking dicts (ext)            0.258662, 0.269982, 0.260755, 0.263075
unpacking dicts (fallback)     11.955657, 12.018308, 11.945628, 11.894115
````
The result with my PR barnch:
```
packing integers (ext)           0.333752, 0.339650, 0.347493, 0.359014
packing integers (fallback)      4.536829, 4.540393, 4.569847, 4.691006
unpacking integers (ext)         0.066953, 0.067022, 0.066793, 0.067799
unpacking integers (fallback)    8.337325, 8.405046, 8.332201, 8.512055
packing bytes (ext)              0.034619, 0.035266, 0.036648, 0.036447
packing bytes (fallback)         0.762590, 0.793024, 0.764875, 0.745429
unpacking bytes (ext)            0.023080, 0.022725, 0.022505, 0.022861
unpacking bytes (fallback)       2.303375, 2.322663, 2.262317, 2.418983
packing lists (ext)              0.711621, 0.721075, 0.726628, 0.727309
packing lists (fallback)       10.523426, 10.661116, 10.569996, 10.661933
unpacking lists (ext)            0.301531, 0.293438, 0.296860, 0.307903
unpacking lists (fallback)     11.514852, 11.516598, 11.434329, 11.441009
packing dicts (ext)              0.416874, 0.415810, 0.425535, 0.435006
packing dicts (fallback)       12.687544, 12.636342, 12.754518, 12.723379
unpacking dicts (ext)            0.260288, 0.256923, 0.256135, 0.258123
unpacking dicts (fallback)     11.887248, 11.940982, 11.871376, 11.797869
```

**Comment by methane** [comment]
OK, would you try the benchmark with PyPy too?
As far as I remember, PyPy does spetial optimization for `struct.unpack()`. I am afraid that moving format from literal to variable makes the optimization doesn't work.

Note that, extension can be used by CPython when performance is needed, but fallback is used by PyPy always.

**Comment by GuyTuval** [comment]
> 
> 
> OK, would you try the benchmark with PyPy too?
> As far as I remember, PyPy does spetial optimization for `struct.unpack()`. I am afraid that moving format from literal to variable makes the optimization doesn't work.
> 
> Note that, extension can be used by CPython when performance is needed, but fallback is used by PyPy always.
I tried on PyPy 2.7.
With master branch:
```
packing integers (fallback)      0.186303, 0.176621, 0.170210, 0.169829
unpacking integers (fallback)    0.078988, 0.058202, 0.057949, 0.058773
packing bytes (fallback)         0.078529, 0.041171, 0.039028, 0.041117
unpacking bytes (fallback)       0.136437, 0.094988, 0.095692, 0.097415
packing lists (fallback)         0.231822, 0.204713, 0.201140, 0.205309
unpacking lists (fallback)       0.512283, 0.486721, 0.487005, 0.491213
packing dicts (fallback)         0.320112, 0.277198, 0.271035, 0.268374
unpacking dicts (fallback)       0.386129, 0.369852, 0.376348, 0.370727
```
with my PR branch:
```
packing integers (fallback)      0.192539, 0.175923, 0.163118, 0.160464
unpacking integers (fallback)    0.078167, 0.057289, 0.058188, 0.059882
packing bytes (fallback)         0.073200, 0.040191, 0.041237, 0.041831
unpacking bytes (fallback)       0.158312, 0.110172, 0.103099, 0.102477
packing lists (fallback)         0.221874, 0.201449, 0.195479, 0.199038
unpacking lists (fallback)       0.527486, 0.506253, 0.490690, 0.529199
packing dicts (fallback)         0.295316, 0.262627, 0.258951, 0.256411
unpacking dicts (fallback)       0.405497, 0.399488, 0.382478, 0.382264
```

**Comment by GuyTuval** [comment]
> 
> 
> I tried with PyPy 2.7.
> With master branch:
> 
> ```
> packing integers (fallback)      0.186303, 0.176621, 0.170210, 0.169829
> unpacking integers (fallback)    0.078988, 0.058202, 0.057949, 0.058773
> packing bytes (fallback)         0.078529, 0.041171, 0.039028, 0.041117
> unpacking bytes (fallback)       0.136437, 0.094988, 0.095692, 0.097415
> packing lists (fallback)         0.231822, 0.204713, 0.201140, 0.205309
> unpacking lists (fallback)       0.512283, 0.486721, 0.487005, 0.491213
> packing dicts (fallback)         0.320112, 0.277198, 0.271035, 0.268374
> unpacking dicts (fallback)       0.386129, 0.369852, 0.376348, 0.370727
> ```
> 
> with my PR branch:
> 
> ```
> packing integers (fallback)      0.192539, 0.175923, 0.163118, 0.160464
> unpacking integers (fallback)    0.078167, 0.057289, 0.058188, 0.059882
> packing bytes (fallback)         0.073200, 0.040191, 0.041237, 0.041831
> unpacking bytes (fallback)       0.158312, 0.110172, 0.103099, 0.102477
> packing lists (fallback)         0.221874, 0.201449, 0.195479, 0.199038
> unpacking lists (fallback)       0.527486, 0.506253, 0.490690, 0.529199
> packing dicts (fallback)         0.295316, 0.262627, 0.258951, 0.256411
> unpacking dicts (fallback)       0.405497, 0.399488, 0.382478, 0.382264
> ```

**Comment by GuyTuval** [comment]
> 
> 
> OK, would you try the benchmark with PyPy too?
> As far as I remember, PyPy does spetial optimization for `struct.unpack()`. I am afraid that moving format from literal to variable makes the optimization doesn't work.
> 
> Note that, extension can be used by CPython when performance is needed, but fallback is used by PyPy always.

I tried with PyPy 2.7.
With master branch:
```
packing integers (fallback)      0.186303, 0.176621, 0.170210, 0.169829
unpacking integers (fallback)    0.078988, 0.058202, 0.057949, 0.058773
packing bytes (fallback)         0.078529, 0.041171, 0.039028, 0.041117
unpacking bytes (fallback)       0.136437, 0.094988, 0.095692, 0.097415
packing lists (fallback)         0.231822, 0.204713, 0.201140, 0.205309
unpacking lists (fallback)       0.512283, 0.486721, 0.487005, 0.491213
packing dicts (fallback)         0.320112, 0.277198, 0.271035, 0.268374
unpacking dicts (fallback)       0.386129, 0.369852, 0.376348, 0.370727
```
with my PR branch:
```
packing integers (fallback)      0.192539, 0.175923, 0.163118, 0.160464
unpacking integers (fallback)    0.078167, 0.057289, 0.058188, 0.059882
packing bytes (fallback)         0.073200, 0.040191, 0.041237, 0.041831
unpacking bytes (fallback)       0.158312, 0.110172, 0.103099, 0.102477
packing lists (fallback)         0.221874, 0.201449, 0.195479, 0.199038
unpacking lists (fallback)       0.527486, 0.506253, 0.490690, 0.529199
packing dicts (fallback)         0.295316, 0.262627, 0.258951, 0.256411
unpacking dicts (fallback)       0.405497, 0.399488, 0.382478, 0.382264
```

**Comment by GuyTuval** [comment]
Hi @methane! Any other checks are needed?

**Comment by GuyTuval** [comment]
Hi @bwesterb! 😃 
I've seen you are one of the contributors to this repo.
Would you mind looking at my PR and merge it in case the PR is okay?

**Comment by methane** [comment]
I don't have time to look it in this year. Please wait months.

**Comment by GuyTuval** [comment]
Hello @methane 😀
Can you check out this PR?

File: msgpack/fallback.py [hunk 2]
```diff
@@ -438,191 +474,82 @@ def _read_header(self, execute=EX_CONSTRUCT):
...
-            if n > self._max_bin_len:
-                raise ValueError("%s exceeds max_bin_len(%s)" % (n, self._max_bin_len))
-            obj = self._read(n)
-        elif b == 0xC6:
-            typ = TYPE_BIN
-            self._reserve(4)
-            n = _unpack_from(">I", self._buffer, self._buff_i)[0]
-            self._buff_i += 4
+        elif 0xC2 <= b <= 0xC3:
+            obj = b == 0xC3
```
Comment (methane) [inline msgpack/fallback.py:478]: revert this.

File: msgpack/fallback.py [hunk 3]
```diff
@@ -149,6 +149,42 @@ def _unpack_from(f, b, o=0):
     _unpack_from = struct.unpack_from
 
 
+MessagePack = namedtuple("MessagePack", ["size", "format", "type"])
+MessagePack.__new__.__defaults__ = ("", -1)
```
Comment (methane) [inline msgpack/fallback.py:153]: I don't like namedtuple here.
Use simple tuple, and `size, fmt, type = _MSGPACK_HEADERS[b]` instead.

File: msgpack/fallback.py [hunk 4]
```diff
@@ -149,6 +149,42 @@ def _unpack_from(f, b, o=0):
...
+
+
+MESSAGE_PACK_DICT = {
```
Comment (methane) [inline msgpack/fallback.py:156]: Don't add public name. Use `_MSGPACK_HEADERS`.

File: msgpack/fallback.py [hunk 5]
```diff
@@ -409,7 +445,7 @@ def _reserve(self, n, raise_outofdata=True):
             self._buff_i = 0  # rollback
             raise OutOfData
 
-    def _read_header(self, execute=EX_CONSTRUCT):
+    def _read_header(self):
```
Comment (methane) [inline msgpack/fallback.py:448]: nice catch.

**Pull request merged**